---
title: "Phylogenetic Tree Construction"
output: 
  html_document:
    toc: yes
    toc_float:
      collapse: no
      smooth_scroll: yes
      toc_depth: 3
    theme: "default"
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", fig.path = "../figures/03_Phylogenetic_Tree") # Send any figures to this folder
```

# Setting the Environment

## Set the seed
```{r set-seed}
set.seed(31415)
```

# Goals of this file

The goal is to create a phylogenetic tree!

1. Load in preprocessed phyloseq object.
2. Create an ASV fasta file from the phyloseq object.
3. Align the 16S sequences from fasta file with MAFFT.
4. Create a tree with FastTree2.

## Load Libraries & Phyloseq object
```{r load-packages}
#install.packages("iNEXT")
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

load("/workdir/zlr6/git_repos/SalinityGradient_16s/data/02_PreProcessing/raw_preprocessed_physeq.RData")
```

# Create fasta file of ASVs and their sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fasta}
# pull out ASV seqs and ASV names
asv_seq_df <- raw_preprocessed_physeq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

# Add the > symbol to make the fasta header
asv_seq_df$ASV <- paste0(">",asv_seq_df$ASV)

# Create the fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseq))

# Write to a fasta
write(asv_seq_fasta, file = "data/03_Phylogenetic_Tree/preprocessed_ASVs.fasta")
```

# Align the 16S sequences from fasta file with MAFFT

```{r run-mafft, engine = 'bash', engine.opts = '-l'}
# Write bash code to run mafft
# Provide the path to mafft
export PATH=/programs/mafft/bin:$PATH

# Change directories to provide the fasta file we made above
cd data/03_Phylogenetic_Tree/
pwd

# Set a seed
RANDOM=31415

# Run Mafft
# For now, use default options
/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_ASVs.fasta

# change back to the project directory
cd ../../
pwd
```

# FastTree2

```{r run-fasttree2, engine = 'bash', engine.opts = "-l"}
# Write bash code to run fasttree2
# Provide the path to fasttree2
export PATH=/programs/FastTree-2.1.11:$PATH

# Change directories to provide the fasta file we made above
cd data/03_Phylogenetic_Tree/
pwd

# Run fasttree2
# Parameters:
  # -nt = indicates it's a nucleotide alignment
  # -gtr = generalized time reversible substitution mode
  # -fastest = speed up the model, reduce memory usage
  # -log = output a log model
  # input alignment file
  # specify the output tree file

FastTree -nt -gtr -log FastTree.log MAFFT_aligned_ASVs.fasta > ASVs_unrooted.tree

# change back to the project directory
cd ../../
pwd
```

------------------------------------------------------

# Session Information
```{r session-info}
devtools::session_info()
```
