---
title: "Between Sample (Beta) Diversity of Microbes along a Salinity Gradient"
output: 
  html_document:
    toc: yes
    toc_float:
      collapse: no
      smooth_scroll: yes
      toc_depth: 3
    theme: "default"
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---
---
title: "Phyolgenetic Tree Inspection & Rooting"
output: 
  html_document:
    toc: yes
    toc_float:
      collapse: no
      smooth_scroll: yes
      toc_depth: 3
    theme: "default"
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", # Always relative to the document directory
                      fig.path = "../figures/05_Community_Analysis") # Send any figures to this folder
```

## Goals

1. Load in phyloseq data with rooted tree.
2. Evaluate sequencing depth and remove sample.
3. Normalize the read counts between samples.
4. Calculate community dissimilarities. (Numbers between 0 and 1. If 0, completely similar versus if 1, completely dissimilar.)
  a. *Sorensen*: Shared species as a binary value: Abundance-unweighted
  b. *Bray-Curtis*: Shared abundant species: Abundance-weighted
  c. *(Abundance) Weighted UNIFRAC*: Consider abundant species and where they fall on the tree
5. Visualize the community data with two unconstrained Ordinations:
  a. *PCoA*: Linear method, calculates eigenvalue = how much variation is explained by each axis. Choose to view axis 1, 2, 3, etc. and plot them together.
  b. *NMDS*: Multiple dimensions into 2 or 3, need to report stress value (ideally <0.15)
6. Run statistics with PERMANOVA and betadispR.

# Setup

## Set Seed

```{r set-seed}
set.seed(31415)
```

## Load Libraries
```{r load-libraries}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan, install = FALSE)
# Load station colors
station_colors <- c(
  "Shipping Channel" = "dodgerblue4",
  "Aransas Bay" = "dodgerblue2",
  "Copano West" = "#D9CC3C",
  "Copano East" = "#A0E0BA",
  "Mesquite Bay" = "#00ADA7")
```

## Load Data
```{r load-data}
# Load in rooted phylogenetic tree!
load("data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
midroot_physeq_rm456
unrooted_physeq_rm456
```

# Explore Read Counts
```{r calc-seq-depth}
# calculate the total number of reads per samplee
raw_TotalSeqs_df <- 
  midroot_physeq_rm456 %>%
  # calculate the sample read sums
  sample_sums() %>%
  data.frame()
# name the column
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"

head(raw_TotalSeqs_df)

# make histogram of raw reads
raw_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0, 10000)) +
  labs(title = "Raw Sequencing Depth Distribution") +
  theme_classic()
```

## Remove lowly seq sample
```{r low-seq-sample}
raw_rooted_physeq <- 
  midroot_physeq_rm456 %>%
  subset_samples(names != "20210615-MA-ABB2F") %>%
  prune_taxa(taxa_sums(.) > 0, .)
# inspect
raw_rooted_physeq

# What is the min number of sequences?
raw_rooted_physeq %>%
  sample_sums() %>%
  min()
```

## Normalize Read Counts
```{r scale-reads-function}
# Scale reads function and also matround function 

### scale_reads function
#################################################################################### 
# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}
```

## Scale the reads and check the distribution of the seq depth
```{r scale-physeq}
min(sample_sums(raw_rooted_physeq))

# scale reads by above functions
scaled_rooted_physeq <- raw_rooted_physeq %>%
  scale_reads(round = "matround")

# calculate the read depth
scaled_TotalSeqs_df <-
  scaled_rooted_physeq %>%
  sample_sums() %>%
  data.frame() 
colnames(scaled_TotalSeqs_df)[1] <- "TotalSeqs"

# Inspect
head(scaled_TotalSeqs_df)

# check the range of the data
min(scaled_TotalSeqs_df)
max(scaled_TotalSeqs_df)

# plot it!
scaled_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0, 10000)) + 
  labs(title = "Scaled Sequencing Depth at 2194") + 
  theme_classic()
```

# Cacluate & Visualize community dissimilarity 

Exploratory analyses from Paliy & Shankar (2016) paper, which is using unconstrained ordination methods like PCoA. 

## PCoA 

### Sorensen
```{r sorensen-PCoA}
# Calculate sorensen dissimilarity: Abundance-unweighted of shared taxa
scaled_soren_pcoa <-  
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray", binary = TRUE)

#str(scaled_soren_pcoa)

# Plot the ordination 
soren_station_pcoa <- plot_ordination(
  physeq = scaled_rooted_physeq,
  ordination = scaled_soren_pcoa,
  color = "station",
  shape = "station",
  title = "Sorensen PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = station)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = station_colors) + 
  theme_bw()
# Show the plot 
soren_station_pcoa
```

*Note that I have removed the PERMANOVA test to below the ordinations. We will come back to it later!*

Here, we are evaluating the shared taxa by presence/absence (abundance-unweighted) in the Sorensen metric. 

Note that there is a weird arch or horseshoe-effect in the data. This is likely because we have complete species turnover between our stations. If you have this in your data, please take a look at the paper by [Morton et al., 2017](https://journals.asm.org/doi/10.1128/msystems.00166-16). More on this in NMDS...

Note that:  

- Axis 1 = ~21% of variation  
- Axis 2 = ~15% of variation 

This means we explain 36% of the variation in the data in these two axes. 



### Bray
```{r bray-PCoA}
# Calculate the BC distance
scaled_BC_pcoa <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray")

# Plot the PCoA
bray_station_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_BC_pcoa,
    color = "station",
    shape = "station",
    title = "Bray-Curtis PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = station)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(station_colors)) + 
  theme_bw()
bray_station_pcoa
```

Here, we are evaluating the shared taxa and then weighting them by their abundances, which provides more influence for species that are more abundant. 

Note that there is a weird arch or horseshoe-effect in the data. This is likely because we have complete species turnover between our stations. If you have this in your data, please take a look at the paper by [Morton et al., 2017](https://journals.asm.org/doi/10.1128/msystems.00166-16). More on this in NMDS...

Note that:  

- Axis 1 = ~28% of variation  
- Axis 2 = ~20% of variation 

This means we explain 48% of the variation in the data in these two axes, which is more than the previous plot with the Sorensen Dissimilarity. Abundance does seem to have an influence!!

It also looks like the samples are now separating more within each group than they did a bit more with Sorensen. Please note this with how the Copano West samples are looking between the Bray-Curtis and the Sorensen plots. 



### Weighted Unifrac
```{r wUnifrac-PCoA}
# Calculate the BC distance
scaled_wUNI_pcoa <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "wunifrac")

# Plot the PCoA
wUNI_station_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_pcoa,
    color = "station",
    shape = "station",
    title = "Weighted Unifrac PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = station)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(station_colors)) + 
  theme_bw()
wUNI_station_pcoa
```

Here, we are evaluating the shared taxa and then weighting them by their abundances, which provides more influence for species that are more abundant. 

Note that there is a weird arch or horseshoe-effect in the data. This is likely because we have complete species turnover between our stations. If you have this in your data, please take a look at the paper by [Morton et al., 2017](https://journals.asm.org/doi/10.1128/msystems.00166-16). More on this in NMDS...

Note that:   

- Axis 1 = ~41% of variation  
- Axis 2 = ~29% of variation 

This means we explain *70*% of the variation in the data in these two axes (!!!), which is significantly more than the previous plots with the taxonomic dissimilarity measures. Here, phylogeny seems to be very important! This means that taxa that are abundant are found in different places in the phylogenetic tree. Therefore, the evoultionary distances (aka the branch lengths) and their abundances seem to have a major influence!!

It also looks like the samples are now separating more within each group than they did a bit more with Sorensen. Please note this with how the Copano West samples are looking between the Bray-Curtis and the Sorensen plots. 


## Combine PCoAs

Let's plot all three together into one plot to have a concise visualization of the three metrics. 
```{r pcoa-together, fig.width=8, fig.height=3.5}
(soren_station_pcoa + theme(legend.position = "none")) + 
  (bray_station_pcoa + theme(legend.position = "none")) + 
    (wUNI_station_pcoa + theme(legend.position = "none"))
```


## NMDS 

### Weighted Unifrac

Since we did 3 of the dissimilarity metrics for the PCoA, let's just plot one example of them for the NMDS plotting. Here, we will use weighted Unifrac 
```{r wUnifrac-NMDS}
# Calculate the Weighted Unifrac distance
scaled_wUNI_nmds <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "NMDS",
    distance = "wunifrac")

# Plot the PCoA
wUNI_station_nmds <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_nmds,
    color = "station",
    shape = "station",
    title = "Weighted Unifrac NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = station)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(station_colors)) + 
  theme_bw()
wUNI_station_nmds
```

We can see from above the plot that the stress value is ~0.15, which is just barely at the limit of the acceptable stress value. And, It seems important to emphasize that the PCoA and the NMDS plot both look pretty similar! 

In this case, I would always prefer to report the PCoA results because they are linear and provide a lot more post-hoc analyses to follow up with. In addition, it's helpful to only have 2 axes of variation and show how much variation is explained. 

```{r combined-wUnifrac, fig.width=6, fig.height=3.5}
(wUNI_station_pcoa + theme(legend.position = "none")) + 
  (wUNI_station_nmds + theme(legend.position = "none"))
```


# Statistical Significance Testing 

## PERMANOVA 
```{r PERMANOVA}
# Calculate all three of the distance matrices
scaled_sorensen_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray", binary = TRUE)
scaled_bray_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray")
scaled_wUnifrac_dist <- phyloseq::distance(scaled_rooted_physeq, method = "wunifrac")

# make a data frame from the sample_data
# All distance matrices will be the same metadata because they 
# originate from the same phyloseq object. 
metadata <- data.frame(sample_data(scaled_rooted_physeq))

# Adonis test
# In this example we are testing the hypothesis that the five stations
# that were collected have different centroids in the ordination space 
# for each of the dissimilarity metrics, we are using a discrete variable 
adonis2(scaled_sorensen_dist ~ station, data = metadata)
adonis2(scaled_bray_dist ~ station, data = metadata)
adonis2(scaled_wUnifrac_dist ~ station, data = metadata)
```

Note that:  

- R2 = the percent variation explained. 
- F = the F-Statistic, which represents the importance value. 
- Pr(>F) = the pvalue 


Above, we see that the most variation is explained by the weighted unifrac, which explains ~45% of the variation in the data and also has the highest F-statistic.



```{r PERMANOVA-multiVariable}
# We might also care about other variables
# Here, we will add date and fraction as variables
# multiplicative model ORDER MATTERS! 
adonis2(scaled_sorensen_dist ~ station * date * fraction, data = metadata)
adonis2(scaled_bray_dist ~ station * date * fraction, data = metadata)
# Note that the ORDER MATTERS!
adonis2(scaled_wUnifrac_dist ~ station * date * fraction, data = metadata)
adonis2(scaled_wUnifrac_dist ~ date * station * fraction, data = metadata)
```

We can also run tests that include additive (+) or multipliciatve models, which include the interaction term between variables. 

## BetaDispR

The PERMANOVA is sensitive to variance/dispersion in the data. Therefore, we need to run a homogeneity of dispersion test to test for the sensitivity of our PERMANOVA results to variance. 
```{r betadispR}
# Homogeneity of Disperson test with beta dispr
# Sorensen 
beta_soren_station <- betadisper(scaled_sorensen_dist, metadata$station)
permutest(beta_soren_station)

# Bray-curtis 
beta_bray_station <- betadisper(scaled_bray_dist, metadata$station)
permutest(beta_bray_station)

# Weighted Unifrac 
beta_bray_station <- betadisper(scaled_wUnifrac_dist, metadata$station)
permutest(beta_bray_station)
```

# Taxanomic Composition

## Phylum 

```{r phylum-colors}
# Set the phylum colors
phylum_colors <- c(
  Acidobacteriota = "navy", 
  Actinobacteriota = "darkslategray2", 
  Armatimonadota = "deeppink1",
  Alphaproteobacteria = "plum2", 
  Bacteroidota = "gold", 
  Betaproteobacteria = "plum1", 
  Bdellovibrionota = "red1",
  Chloroflexi="black", 
  Crenarchaeota = "firebrick",
  Cyanobacteria = "limegreen",
  Deltaproteobacteria = "grey", 
  Desulfobacterota="magenta",
  Firmicutes = "#3E9B96",
  Gammaproteobacteria = "greenyellow",
  "Marinimicrobia (SAR406 clade)" = "yellow",
  Myxococcota = "#B5D6AA",
  Nitrospirota = "palevioletred1",
  Proteobacteria = "royalblue",
  Planctomycetota = "darkorange", 
  "SAR324 clade(Marine group B)" = "olivedrab",
  #Proteobacteria_unclassified = "greenyellow",
  Thermoplasmatota = "green",
  Verrucomicrobiota = "darkorchid1")
 # Other = "grey")

```

### Plot Phylum Composition

```{r phylum-composition}
# calculate the phylum relative abundance
# Note: the read depth must be normalized in some way: sale_reads
phylum_df <- scaled_rooted_physeq %>%
  # agglomerate at the phylum table
  tax_glom(taxrank = "Phylum") %>%
  # Transform counts to relative abundance
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format
  psmelt() %>%
  # filter out phyla that are less than 1% - get rid of low abundant phyla
  dplyr::filter(Abundance > 0.01) %>%
  # fix the order of the date
  mutate(date = fct_relevel(date, c("6/2/21", "6/15/21", "10/5/21")),
         station = fct_relevel(station, c("Copano West", "Copano East", "Mesquite Bay", "Aransas Bay", "Shipping Channel")))

# plot phylum abunsances - make sure to load phylum colors
phylum_df %>%
  # It's important to have one sample per x value
  # otherwise, it will take the sum between multiple samples
  dplyr::filter(depth == 0.0) %>%
  dplyr::filter(fraction == "Whole") %>%
  ggplot(aes(x = station, y = Abundance, fill= Phylum)) +
  facet_grid(Phylum~date, scale = "free") +
  geom_bar(stat = "identity", color = "black") +
  labs(Title = "Surface Phylum Composition") +
  scale_fill_manual(values = phylum_colors) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

# Narrow in on a specific group
# Actinobacteriota - y: abundance, x: station, dot plot + boxplot
phylum_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # build the plot 
  ggplot(aes(x = station, y = Abundance, fill = station, color = station)) +
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + #outliers not plotted here in boxplot
  geom_jitter() +
  theme_bw() +
  labs(title = "Actinobacteriota Phylum Abundance") + 
  scale_color_manual(values = station_colors) +
  scale_fill_manual(values = station_colors)
  
# Statistics: Kruskall-Wallis folowed by a Tukey's Posthoc tets
# These are non-parametric for non-normal stat tests
```

##Family

### Plot Family Composition

```{r family-composition}
# calculate the family relative abundance
# Note: the read depth must be normalized in some way: sale_reads
family_df <- scaled_rooted_physeq %>%
  # agglomerate at the phylum table
  tax_glom(taxrank = "Family") %>%
  # Transform counts to relative abundance
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format
  psmelt() %>%
  # filter out phyla that are less than 1% - get rid of low abundant phyla
  dplyr::filter(Abundance > 0.01) %>%
  # fix the order of the date
  mutate(date = fct_relevel(date, c("6/2/21", "6/15/21", "10/5/21")),
         station = fct_relevel(station, c("Copano West", "Copano East", "Mesquite Bay", "Aransas Bay", "Shipping Channel")))

# Narrow in on a specific group
# Actinobacteriota - y: abundance, x: station, dot plot + boxplot
family_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # build the plot 
  ggplot(aes(x = station, y = Abundance, fill = station, color = station)) +
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + #outliers not plotted here in boxplot
  geom_jitter() +
  theme_bw() +
  labs(title = "Actinobacteriota Family Abundance") + 
  scale_color_manual(values = station_colors) +
  scale_fill_manual(values = station_colors)
  
# Statistics: Kruskall-Wallis folowed by a Tukey's Posthoc tets
# These are non-parametric for non-normal stat tests
```

---

# Session Information
```{r session-info}
devtools::session_info()
```
